Fixes players being unable to construct on the same tile as a directional window

Irafas opened the pull request:
:cl:
fix: players can construct while standing on top of a directional window. (machine frames, computer frames, etc)
/:cl:

fixes: #39078 

Users could drag stuff onto the same tile as a directional window and anchor it down, but couldn't build them while standing on the tile. This adds directional windows to the list of objects that construction is allowed on top of.

Irafas committed:
Can build on top of windows

-Players can now build things while standing on the same tile as a directional window.

Irafas committed:
Changed the check for fulltile

-This actually covers all fulltile windows

Cyprex commented:
Will this allow you to stack directional windows? Having several windows on top of each other would be pretty bad.

Irafas commented:
@Cyprex Nah a separate check prevents directional windows from being built on top of each other. Though when I just tested this now I've found that a fulltile window can be built on the same tile as a directional window if the player faces a direction where no directional windows are. So while windows can't be stacked multiple times you can additionally reinforce a fulltile window with a single layer of directional windows

Irafas committed:
No more stacking windows

-Full tiles can no longer be built on top of directional windows
-Directional windows can still be built on the same tile as long as they dont face the same direction

ExcessiveUseOfCobblestone commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -246,6 +246,13 @@
+				var/obj/structure/window/W = AM
+				var/obj/structure/window/O = new R.result_type
+				if(istype(O,/obj/structure/window) && O.fulltile)
+					//Do nuthin and move on
```
really...

Irafas replied:
looking at this I realize that I should have compressed them into one if. Doing that now

Irafas committed:
Now contains 50% less IF!!

-Put both if's into one

Dennok commented:
Check if directional window can be rotated to another directional window direction. If yes you steel can stack.

AnturK commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -246,6 +246,11 @@
 				continue
+			if(istype(AM,/obj/structure/window))
+				var/obj/structure/window/W = AM
+				var/obj/structure/window/O = new R.result_type
```
What are you trying to do here, you're creating random objects here in nullspace on every build check.

AnturK commented:
Assuming you want to prevent fulltile windows being built on top of directionals with that the real issue is invalid direction being passed in valid_window_location above. You need to fix that instead.

Irafas commented:
@AnturK I've been looking at the valid_window_location part and I can't see what you mean by an invalid direction. The proc gets passed the users turf and their direction but not the object they're trying to construct. Even if it did I would still have to instantiate it so I could check if the var `fulltile` is true like I've already done. 

I want to keep the behavior of what I currently have where the user can build multiple directional windows on the same tile as long as they aren't on the same side. Without checking if the structure being built is a fulltile window I can't prevent them from building one on top of those directional windows.

I could just use a `qdel(O)` to delete the object I made for finding out if the being constructed object is a fulltile window. 

Thanks for pointing out the nullspace issue though. I messed up and didn't realize that creating a new object doesn't mean they fall out of scope like variables that only point toward a different object.

AnturK commented:
The wrong part is the passed the direction, the direction the function expects is the direction of the window. Which for fulltile windows is FULLTILE_WINDOW_DIR not the current user's direction.

Irafas committed:
Still pretty fucky

-If the dir of the constructed object has var fulltile set to TRUE then FULLTILE_WINDOW_DIR will be passed to the method valid_window_location else the user's dir is passed.

Irafas committed:
just a space

Irafas commented:
So to get the dir of the object to be constructed I have to first make it. Is there a way to get the variables of an object out of the recipe without first having to construct it?

optimumtact commented:
if you have a reference to the type of an object you can do
initial(type.variablename) and get the compile time value of that variable, this works for most basic variables but not lists and some other meme crap.

is that what you were after?

Irafas commented:
@optimumtact Yeah actually that was exactly what I was looking for. I'll see if I can get it to work with this now. Thanks for the help.

Irafas committed:
Now gets initial dir of object being built and passes that to valid_window_location

ExcessiveUseOfCobblestone commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -231,7 +231,9 @@
 	var/turf/T = get_turf(usr)
-	if(R.window_checks && !valid_window_location(T, usr.dir))
+
+	var/obj/D = R.result_type
```
This isn't an instance can you just type it as `var/D`

Irafas replied:
Actually the `var/obj/D` is needed as else it doesn't compile due to the fact that the dir var is in the /obj/ type.

ExcessiveUseOfCobblestone replied:
@AnturK 

That doesn't sound right.

You can do `initial(D.dir)` but you can't typecast it as an instance because it's a type. You're just tricking the compiler into compiling.

AnturK replied:
I have no idea what you're talking about this is exactly how you use initial on types.

ExcessiveUseOfCobblestone replied:
I have been told not to typecast things as instances (i think for confusion to others who are reading might think it's an actual item), but this works(tm) since it's doing `initial(D.blahblah)` in the later blocks.

Otherwise it would cause runtimes, it might be better to just put a pin on this and make it clear that's a fake object whenever this code gets touched again.

Irafas replied:
unfortunately `initial(R.result_type.dir)` also doesn't work. The compiler doesn't know that the object has a dir variable. I hoped that I could do that at first but ¯\\\_(ツ)_/¯

ExcessiveUseOfCobblestone replied:
please be slow I have problems.

Thanks for explaining, sucks tho.

ExcessiveUseOfCobblestone commented:
oh lol it was merged.
